### PR TITLE
feat(cdk): add async job components

### PR DESCRIPTION
**Pull request by maxi297**
## What
Creating an async job retriever component and adding it to the declarative manifest.

## How
The design has been described [here](https://whimsical.com/cdk-async-jobs-XcTdHYRBHzKVFK2hDc3iQn@2bsEvpTYSt1HjDYTYmK5i8sNMpoNEVwQdVN). 

We have three main parts:
* A retriever which interfaces with the DeclarativeStream/StreamSlicer and does filtering/transformation
* A JobOrchestrator which handles the logic associated to jobs
* A HttpJobRepository which handles HTTP communications

The usage can be seen [here](https://github.com/airbytehq/airbyte/compare/async-job/feature-branch#diff-b041c21db61fd5fe7fb67481e6167b6a343236571089c1ed61498b597202a750R879-R928)

Also part of this PR is adding the addition of a transformation for sendgrid and fixing the interface. This has already been reviewed [here](https://github.com/airbytehq/airbyte/pull/45135)

## User Impact
This will allow us to update source-sendgrid to be manifest only source.

## Can this PR be safely reverted and rolled back?

- [X] YES 💚
- [ ] NO ❌

### CODE REVIEW

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/http_job_repository.py [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+        Args:
+            job (AsyncJob): The job to fetch records from.
+
+        Yields:
+            Iterable[Mapping[str, Any]]: A generator that yields records as dictionaries.
+
+        """
+
+        for url in self.urls_extractor.extract_records(self._polling_job_response_by_id[job.api_job_id()]):
+            stream_slice: StreamSlice = StreamSlice(partition={"url": url}, cursor_slice={})
```
Comment (maxi297) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/http_job_repository.py:155]: Because we don't have interpolation on random fields, we have this hack which consist on using the stream_slice to allow for interpolation. This is what it looks like in the manifest: https://github.com/airbytehq/airbyte/compare/async-job/feature-branch#diff-b041c21db61fd5fe7fb67481e6167b6a343236571089c1ed61498b597202a750R921

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/http_job_repository.py [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+
+        Args:
+            jobs (Iterable[AsyncJob]): An iterable of AsyncJob objects representing the jobs to update.
+
+        Returns:
+            None
+        """
+        for job in jobs:
+            stream_slice = StreamSlice(
+                partition={"create_job_response": self._create_job_response_by_id[job.api_job_id()]},
```
Comment (maxi297) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/http_job_repository.py:133]: Because we don't have interpolation on random fields, we have this hack which consist on using the stream_slice to allow for interpolation. This is what it looks like in the manifest: https://github.com/airbytehq/airbyte/compare/async-job/feature-branch#diff-b041c21db61fd5fe7fb67481e6167b6a343236571089c1ed61498b597202a750R913

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/extractors/response_to_file_extractor.py [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+
+        content_type = headers.get("content-type")
+
+        if not content_type:
+            return DEFAULT_ENCODING
+
+        content_type, params = requests.utils.parse_header_links(content_type)
+
+        if "charset" in params:
+            return params["charset"].strip("'\"")  # type: ignore  # we assume headers are returned as str
```
[the remark this answers is not in the document]
Comment (maxi297) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/extractors/response_to_file_extractor.py:52]: Ignoring this one as this is copy/pasted from requests

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/declarative_component_schema.yaml [hunk 4]
```diff
@@ -2341,6 +2356,104 @@ definitions:
...
+        description: Component that describes how to extract records from a HTTP response.
+        "$ref": "#/definitions/RecordSelector"
+      status_mapping:
+        description: Async Job Status to Airbyte CDK Async Job Status mapping.
+        anyOf:
+          - "$ref": "#/definitions/AsyncJobStatusMap"
+      status_extractor:
+        description: Responsible for fetching the actual status of the async job.
+        anyOf:
+          - "$ref": "#/definitions/CustomRecordExtractor"
```
Comment (maxi297) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/declarative_component_schema.yaml:2411]: It is a bit annoying that if we add an extractor, we need to update all the places where we reference extractors. Same thing with the requesters, decoders, etc... Do we have a solution for that?

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/extractors/response_to_file_extractor.py [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+        decompressor = zlib.decompressobj(zlib.MAX_WBITS | 32)
+
+        if response:
+            tmp_file = str(uuid.uuid4())
+            with closing(response) as response, open(tmp_file, "wb") as data_file:
+                response_encoding = self._get_response_encoding(dict(response.headers or {}))
+                for chunk in response.iter_content(chunk_size=DOWNLOAD_CHUNK_SIZE):
+                    try:
+                        data_file.write(decompressor.decompress(chunk))
+                    except zlib.error:
```
[the remark this answers is not in the document]
Comment (maxi297) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/extractors/response_to_file_extractor.py:98]: It seems like codeflash is proposing to remove writting chunks to the file and I don't get the indentation.

I would be tempted to say: Go home codeflash, you are drunk

**Comment by maxi297** [comment]
I know there is still mypy issues but I don't think it is blocking a review

File: airbyte-cdk/python/pyproject.toml [hunk 6]
```diff
@@ -35,6 +35,7 @@ isodate = "~0.6.1"
 Jinja2 = "~3.1.2"
 jsonref = "~0.2"
 jsonschema = "~3.2.0"
+pandas = "2.2.0"
```
Comment (maxi297) [inline airbyte-cdk/python/pyproject.toml:38]: This change causes the smoke tests on shopify and zendesk-support not to run. This is unfortunate. @alafanechere @clnoll is there a way to have the --use-local-cdk flag and make sure all the dependencies are resolved? [This](https://github.com/airbytehq/airbyte/blob/04f40f903c1e901748d7dcdbb753188346a6e2a9/.github/workflows/cdk_connectors_tests.yml#L95) is the command that is failing right now

Comment (alafanechere) [reply airbyte-cdk/python/pyproject.toml:38]: The error could indeed indicate that the connector is not built with the local CDK.
The logic to install the local CDK is handled [here](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/pipelines/dagger/actions/python/common.py#L240).
I checked the dagger logs of your run and it looks like its correctly called.
There could be a clash between where the local CDK is installed via pip and where the connector is installed via poetry (we're not using venv on build so I'd not expect this)...
Now that the CDK is on poetry I think it'd be better to install the local CDK via `poetry` instead of `pip`.
TLDR; we should investigate if the local cdk is really installed when the `--use-local-cdk` flag is passed.

Comment (girarda) [reply airbyte-cdk/python/pyproject.toml:38]: [GH issue](https://github.com/airbytehq/airbyte-internal-issues/issues/8494). CC @natikgadzhi this is preventing us from testing CDK changes on connectors

Comment (natikgadzhi) [reply airbyte-cdk/python/pyproject.toml:38]: Sounds like a high priority regression testing bug to me. I'll spend a week looking into them myself when I'm back, and will see if I can snatch a week of Augustin's time too.

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/declarative_component_schema.yaml (hunk 4 above)
Comment (brianjlai) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/declarative_component_schema.yaml:2411]: unfortunately not to my knowledge :/

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/http_job_repository.py [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+            AsyncJobStatus: The validated job status.
+
+        Raises:
+            ValueError: If the API status is unknown.
+        """
+
+        api_status = next(iter(self.status_extractor.extract_records(response)), None)
+        job_status = self.status_mapping.get(str(api_status), None)
+        if job_status is None:
+            raise ValueError(
```
Comment (maxi297) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/http_job_repository.py:75]: I'm wondering if this is dangerous: what happens if the API adds a status? I guess we should know in some way but Salesforce today only operate on a COMPLETE and FAILED statuses and probably assume the rest is running (see [this](https://github.com/airbytehq/airbyte/blob/f23ae56693c931e36028a61ea1b11ffdd25762e2/airbyte-integrations/connectors/source-salesforce/source_salesforce/streams.py#L421))

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/retrievers/async_retriever.py [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+            )
+        return stream_slice["partition"]  # type: ignore  # stream_slice["partition"] has been added as an AsyncPartition as part of stream_slices
+
+    def stream_slices(self) -> Iterable[Optional[StreamSlice]]:
+        slices = self.stream_slicer.stream_slices()
+        self.__job_orchestrator = self._job_orchestrator_factory(slices)
+
+        for completed_partition in self._job_orchestrator.create_and_get_completed_partitions():
+            yield StreamSlice(
+                partition=dict(completed_partition.stream_slice.partition) | {"partition": completed_partition},
```
Comment (maxi297) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/retrievers/async_retriever.py:94]: @brianjlai I thought this might be interesting for you. This doesn't work very well for concurrent because:
* The slice is not json serializable as it is of object `StreamSlice` and it'll fail [here](https://github.com/airbytehq/airbyte/blob/f23ae56693c931e36028a61ea1b11ffdd25762e2/airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/adapters.py#L276)
* Even if we were to cast it as a dict, this would fail because `completed_partition` is of type `AsyncJob` which is also not serializable

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/async_job/job.py [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+
+from airbyte_cdk import StreamSlice
+from airbyte_cdk.sources.declarative.async_job.timer import Timer
+
+from .status import AsyncJobStatus
+
+
+class AsyncJob:
+    """
+    Note that the timer will only stop once `update_status` is called so the job might be completed on the API side but until we query for
```
Comment (girarda) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/async_job/job.py:15]: can you add a description of what the class's purpose is in addition to the note?

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/async_job/job.py [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+        return self._status
+
+    def job_parameters(self) -> StreamSlice:
+        return self._job_parameters
+
+    def update_status(self, status: AsyncJobStatus) -> None:
+        if self._status != AsyncJobStatus.RUNNING and status == AsyncJobStatus.RUNNING:
+            self._timer.start()
+        elif status in [AsyncJobStatus.FAILED, AsyncJobStatus.TIMED_OUT, AsyncJobStatus.COMPLETED]:
+            self._timer.stop()
```
Comment (girarda) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/async_job/job.py:43]: do we need an else to make sure we handle new values that might be added in the future?

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/async_job/job_orchestrator.py [hunk 11]
```diff
@@ -0,0 +1,239 @@
...
+        failed_status_jobs = (AsyncJobStatus.FAILED, AsyncJobStatus.TIMED_OUT)
+        jobs_to_replace = [job for job in partition.jobs if job.status() in failed_status_jobs]
+        for job in jobs_to_replace:
+            new_job = self._job_repository.start(job.job_parameters())
+            partition.replace_job(job, [new_job])
+
+    def _start_jobs(self) -> None:
+        """
+        Start the jobs for each slice in the slice iterator.
+        This method iterates over the slice iterator and starts a job for each slice.
```
Comment (girarda) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/async_job/job_orchestrator.py:98]: the method also replaces jobs that have failed

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/async_job/job_orchestrator.py [hunk 12]
```diff
@@ -0,0 +1,239 @@
...
+            # update the status only if there are RUNNING jobs
+            self._job_repository.update_jobs_status(running_jobs)
+
+    def _wait_on_status_update(self) -> None:
+        """
+        Waits for a specified amount of time between status updates.
+
+
+        This method is used to introduce a delay between status updates in order to avoid excessive polling.
+        The duration of the delay is determined by the value of `_WAIT_TIME_BETWEEN_STATUS_UPDATE_IN_SECONDS`.
```
Comment (girarda) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/async_job/job_orchestrator.py:137]: is there a reason not to make the wait time configurable?

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/async_job/job_orchestrator.py [hunk 13]
```diff
@@ -0,0 +1,239 @@
...
+        self._debug_log(f"Polling status in progress. There are currently {len(self._running_partitions)} running partitions.")
+
+        # wait only when there are running partitions
+        if self._running_partitions:
+            self._debug_log(f"Waiting for {self._WAIT_TIME_BETWEEN_STATUS_UPDATE_IN_SECONDS} seconds before next poll...")
+            time.sleep(self._WAIT_TIME_BETWEEN_STATUS_UPDATE_IN_SECONDS)
+
+    def _debug_log(self, message: str) -> None:
+        # if statement in order to avoid string formatting if we're not in debug mode
+        if LOGGER.isEnabledFor(logging.DEBUG):
```
Comment (girarda) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/async_job/job_orchestrator.py:151]: I feel like this is something we should do every time we want to log debug messages. Should we wrap this in an abstraction?

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/async_job/job_orchestrator.py [hunk 14]
```diff
@@ -0,0 +1,239 @@
...
+            LOGGER.debug(message)
+
+    def _process_completed_partition(self, partition: AsyncPartition) -> None:
+        """
+        Process a completed partition.
+        Args:
+            partition (AsyncPartition): The completed partition to process.
+        Returns:
+            AsyncPartition: The processed partition.
```
Comment (girarda) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/async_job/job_orchestrator.py:160]: the method doesn't return anything

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/async_job/job_orchestrator.py [hunk 15]
```diff
@@ -0,0 +1,239 @@
...
+            partition (AsyncPartition): The partition to process.
+        Returns:
+            AirbyteTracedException: An exception indicating that at least one job could not be completed.
+        Raises:
+            AirbyteTracedException: If at least one job could not be completed.
+        """
+        status_by_job_id = {job.api_job_id(): job.status() for job in partition.jobs}
+        raise AirbyteTracedException(
+            message=f"At least one job could not be completed. Job statuses were: {status_by_job_id}",
+            failure_type=FailureType.system_error,
```
Comment (girarda) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/async_job/job_orchestrator.py:205]: are all jobs failure indicative of a system error?

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/async_job/repository.py [hunk 16]
```diff
@@ -0,0 +1,23 @@
...
+
+from abc import abstractmethod
+from dataclasses import dataclass
+from typing import Any, Iterable, Mapping, Set
+
+from airbyte_cdk import StreamSlice
+from airbyte_cdk.sources.declarative.async_job.job import AsyncJob
+
+
+@dataclass
```
Comment (girarda) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/async_job/repository.py:11]: why is the interface a dataclass?

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/declarative_component_schema.yaml (hunk 4 above)
Comment (girarda) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/declarative_component_schema.yaml:2411]: can you create a reference to the anyOf?

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/declarative_component_schema.yaml [hunk 17]
```diff
@@ -1685,6 +1687,19 @@ definitions:
       type:
         type: string
         enum: [JsonlDecoder]
+  KeysToLower:
+    title: Keys to Lower Fields
```
Comment (girarda) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/declarative_component_schema.yaml:1691]:
```suggestion
    title: Keys to Lower Case
```
?

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/extractors/response_to_file_extractor.py [hunk 18]
```diff
@@ -0,0 +1,162 @@
...
+DOWNLOAD_CHUNK_SIZE: int = 1024 * 1024 * 10
+
+
+class ResponseToFileExtractor(RecordExtractor):
+    """
+    This class is used when having very big HTTP responses (usually streamed) which would require too much memory so we use disk space as
+    a tradeoff.
+
+    Eventually, we want to support multiple file type by re-using the file based CDK parsers if possible. However, the lift is too high for
+    a first iteration so we will only support CSV parsing using pandas as salesforce and sendgrid were doing.
```
Comment (girarda) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/extractors/response_to_file_extractor.py:27]: do you have a plan for reconciling the file-based parsers with low-code?

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/parsers/model_to_component_factory.py [hunk 19]
```diff
@@ -1162,6 +1175,81 @@ def create_simple_retriever(
...
+                if status in api_status_to_cdk_status:
+                    raise ValueError(
+                        f"API status {status} is already set for CDK status {cdk_status}. Please ensure API statuses are only provided once"
+                    )
+                api_status_to_cdk_status[status] = self._get_async_job_status(cdk_status)
+        return api_status_to_cdk_status
+
+    def _get_async_job_status(self, status: str) -> AsyncJobStatus:
+        match status:
+            case "running":
```
Comment (girarda) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/parsers/model_to_component_factory.py:1197]: should these constants be defined anywhere?

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/parsers/model_to_component_factory.py [hunk 20]
```diff
@@ -1162,6 +1175,81 @@ def create_simple_retriever(
...
+        decoder = self._create_component_from_model(model=model.decoder, config=config) if model.decoder else JsonDecoder(parameters={})
+        record_selector = self._create_component_from_model(
+            model=model.record_selector,
+            config=config,
+            decoder=decoder,
+            transformations=transformations,
+            client_side_incremental_sync=client_side_incremental_sync,
+        )
+        stream_slicer = stream_slicer or SinglePartitionRouter(parameters={})
+        creation_requester = self._create_component_from_model(model=model.creation_requester, decoder=decoder, config=config, name=name)
```
Comment (girarda) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/parsers/model_to_component_factory.py:1230]: would it help debugging if the requesters all had different names? eg `f"{name}_creation"`

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/http_job_repository.py (hunk 7 above)
Comment (girarda) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/http_job_repository.py:75]: yes, this will fail eventually, but I think the desired behavior is indeed to fail loudly if an unexpected status shows up. Happy to revisit the decision if this becomes too noisy and isn't actionable.

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/http_job_repository.py (hunk 2 above)
Comment (girarda) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/http_job_repository.py:133]: maybe worth leaving this comment in the code?

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/http_job_repository.py (hunk 1 above)
Comment (girarda) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/http_job_repository.py:155]: Q: would it make sense to create a new type of slice or partition since this isn't exactly a StreamSlice in the typical sense?

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/retrievers/async_retriever.py [hunk 21]
```diff
@@ -0,0 +1,113 @@
...
+        self._job_orchestrator_factory = self.job_orchestrator_factory
+        self.__job_orchestrator: Optional[AsyncJobOrchestrator] = None
+        self._parameters = parameters
+
+    @property
+    def state(self) -> StreamState:
+        """
+        As a first iteration for sendgrid, there is no state to be managed
+        """
+        return {}
```
Comment (girarda) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/retrievers/async_retriever.py:37]: Q: is this a good time to pull the state management out of the retriever into the DeclarativeStream?

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/retrievers/async_retriever.py (hunk 8 above)
Comment (girarda) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/retrievers/async_retriever.py:94]: is there a real blocker to making `AsyncJob` serializable?

maybe hot take: partitions should be serializable else there's no way to pass the context from a thread to another

**Review by girarda — APPROVED** [review]
this is great! Left a few questions, but nothing blocking on my end

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/async_job/job_orchestrator.py (hunk 13 above)
Comment (maxi297) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/async_job/job_orchestrator.py:151]: Actually, what I did here does not work. The input parameter should be a callable to really avoid string formatting. 

I'm kind of sad that Python doesn't have that by default to be honest. This seems like very straightforward. I can add this in a utils methods that would take a logger and a callable.

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/async_job/job_orchestrator.py (hunk 15 above)
Comment (maxi297) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/async_job/job_orchestrator.py:205]: Today:
* we fail for sendgrid: (see [this](https://github.com/airbytehq/airbyte/blob/58989b04f5542ebd9d245860b4d9ca062f2a5630/airbyte-integrations/connectors/source-sendgrid/source_sendgrid/streams.py#L76-L77))
* we switch to REST for Salesforce (see [this](https://github.com/airbytehq/airbyte/blob/58989b04f5542ebd9d245860b4d9ca062f2a5630/airbyte-integrations/connectors/source-salesforce/source_salesforce/streams.py#L616-L630)) which will be handled by the error handler and the "composite" stream that will wrap the DeclarativeStream and the SalesforceRestStream

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/async_job/repository.py (hunk 16 above)
Comment (maxi297) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/async_job/repository.py:11]: I'll remove that. I guess this was copied from other instances where we were doing this like [here](https://github.com/airbytehq/airbyte/pull/45178/files/07b24c2971c0190ca2462f3b2f1424f132e200bb#diff-74b0a729c5ee94162418441cea80512a9d27920954ed1448547d0334ca2f62aeL14)

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/extractors/response_to_file_extractor.py (hunk 18 above)
Comment (maxi297) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/extractors/response_to_file_extractor.py:27]: Not yet at all sources that could be declarative that I've checked (amazon-seller-partner, salesforce, sendgrid) are reading CSV files. Hence, this helps us push back the decision. In any case, I'm confident in the interfaces in the async jobs Python code and declarative interface (excluding the interpolation) so I'm not too worried

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/parsers/model_to_component_factory.py (hunk 19 above)
Comment (maxi297) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/parsers/model_to_component_factory.py:1197]: I don't see them used anywhere else as they are part of the yaml interface and this is where we do the conversion. Once we use them elsewhere, we might know more about where to extract them

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/parsers/model_to_component_factory.py (hunk 20 above)
Comment (maxi297) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/parsers/model_to_component_factory.py:1230]: Sure! I'll add the prefix for the three requesters

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/http_job_repository.py (hunk 7 above)
Comment (maxi297) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/http_job_repository.py:75]: I'm afraid of this creating P0s as we don't control the API but I also get your point. Let's try and learn from that

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/http_job_repository.py (hunk 1 above)
Comment (maxi297) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/http_job_repository.py:155]: The answer for me would be the allow for interpolation for more parameters than the stream_slice, stream_state and next_page_token as [we limit the interpolation to these](https://github.com/airbytehq/airbyte/blob/58989b04f5542ebd9d245860b4d9ca062f2a5630/airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/request_options/interpolated_request_input_provider.py#L51). This is can be done but will require a change to those three layers: HttpRequester, RequestOptionProvider, RequestInputProvider

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/retrievers/async_retriever.py (hunk 21 above)
Comment (maxi297) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/retrievers/async_retriever.py:37]: Given the deadlines we have for async project, I would say this is out of scope but we can have the discussion

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/retrievers/async_retriever.py (hunk 8 above)
Comment (maxi297) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/retrievers/async_retriever.py:94]: My understanding of how the `json` library works is that we would need to provide a `default`

[doc](https://docs.python.org/3/library/json.html#json.dump) says: `If specified, default should be a function that gets called for objects that can’t otherwise be serialized. It should return a JSON encodable version of the object or raise a [TypeError](https://docs.python.org/3/library/exceptions.html#TypeError). If not specified, [TypeError](https://docs.python.org/3/library/exceptions.html#TypeError) is raised.`

In our case, I don't know if we could:
* assume that if we can't serialize it with the standard encoder, we use the string version
* ask users that puts an object in the slice to implement a `__stream_slice_serialize__` method

I think I would prefer the second option but haven't put too much thoughts into it yet

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/extractors/response_to_file_extractor.py (hunk 18 above)
Comment (girarda) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/extractors/response_to_file_extractor.py:27]: imo there is value in reconciling the two so we don't need to maintain two csv parsers. This should probably be part of the Q3B follow up

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/retrievers/async_retriever.py (hunk 21 above)
Comment (girarda) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/retrievers/async_retriever.py:36]: making the consequence of this decision explicit: async retriever only supports full refresh streams. is that right?

File: airbyte-cdk/python/pyproject.toml (hunk 6 above)
Comment (maxi297) [reply airbyte-cdk/python/pyproject.toml:38]: You can ignore this. This was the issue: https://github.com/airbytehq/airbyte/pull/45362
